design/design-doc-composition-validating-webhook.md

```diff
@@ -0,0 +1,359 @@
+    1. have a valid path as a `fromFieldPath` according to the schema of the
+       source object, if defined for the type of patch and required by the
+       `PatchPolicy` defined.
+    2. have a valid path as a `toFieldPath` according to the schema of the
+       destination object, if defined for the type of patch.
+    3. have correctly configured types, e.g. if `fromFieldPath` is a string,
+       `toFieldPath` is an integer, the patch should be rejected, but not if
+       there is a mapping correctly setting it.
+3. the result of all the `patches` being applied to the `base` should be valid
+   according to the Managed Resource schema.
+
+Compositions are also defining the following fields that we should consider
+validating:
+
+- `patchSets`: a list of named patches that can be referenced by resources.
+  Depending on the type of patch it could be impossible to validate them
+  independently from the resource they would be used on. We could think of only
+  validating them once they are dereferenced and associated to a specific
+  resource.
+- `functions`: a list of Composition Functions that can be used to generate or
+  modify the Composed Resources. Given that it’s an `alpha` feature, we can
+  think of not covering this in the first implementation, however we should
+  keep in mind that having at least a function defined would invalidate the
+  assumption that rendered Managed Resources should be valid according to the
+  schema. Moreover, taking them into consideration during the validation phase
+  would mean adding another external dependency, which could both reduce
+  reliability and add latency.
+- `environment`: contains a list of `EnvironmentConfigs` that can be referenced
+  by specific types of patches and a list of`EnvironmentPatches` . Also an
+  `alpha` feature, similar implications to what already said for `functions`,
+  would require additional information and can invalidate the assumption that
+  we render valid Composed Resources if defined.
+- `writeConnectionSecretsToNamespace` : the name of the namespace the `Secret`
+  containing the connection details should be created into. We could check
+  whether it exists already, but that would be against the behaviour Kubernetes
+  users are used to, e.g. a `Deployment` mounting a `Secret` which does not
+  exist yet is considered valid but remains pending until it’s not created.
+- `publishConnectionDetailsWithStoreConfigRef` : a reference by name to a
+  `StoreConfig`, we could check that it exists, but similarly to the namespace
+  it would go against the behaviour Kubernetes users might be expecting.
+
+### Additional Constraints for the ValidatingWebhook use case
+
+A few constraints that makes validating Compositions from a Validating Webhook
+harder w.r.t. the cases mentioned in the Background:
+
+1. Managed Resources defined as `base` and traversed by `patches` might not yet
+   have been deployed when the Composition is being validated.
```

Comment:
Having an MR always implies having its CRD deployed, but not having MRs does not give any hint about its CRDs existence. So, it would be more clear if we mentioned what we need.